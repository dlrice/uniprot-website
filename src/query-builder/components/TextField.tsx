import React, { useEffect, useState } from 'react';

import { QueryBit, SearchTermType } from '../types/searchTypes';

const getStringValue = (value: string, prefix?: string) =>
  value.includes(' ')
    ? `"${prefix ? `${prefix}` : ''}${value?.trim()}"`
    : `${prefix ? `${prefix}` : ''}${value?.trim()}`;

const TextField: React.FC<{
  field: SearchTermType;
  type: string;
  initialValue?: string;
  handleChange: (queryBit: QueryBit) => void;
}> = ({ field, type, handleChange, initialValue = '' }) => {
  const [value, setValue] = useState(initialValue);

  useEffect(() => {
    if (value.length > 0) {
      if (field.term === 'All') {
        handleChange({
          [field.id]: value?.trim(),
        });
      } else if (
        field.valuePrefix &&
        field.term === 'database' &&
        value.trim() === '*'
      ) {
        // Query is faster with this hack
        handleChange({
          [field.id]: `(${field.term}:${field.valuePrefix})`,
        });
      } else {
        handleChange({
          [field.id]: `(${field.term}:${getStringValue(
            value,
            field.valuePrefix
          )})`,
        });
      }
    }
  }, [field, value, handleChange]);

  return (
    <label htmlFor={`input_${field.term}`}>
      {field.label}
      <input
        id={`input_${field.term}`}
        type={type}
        value={value}
        onChange={(e) => setValue(e.target.value)}
        placeholder={field.example}
      />
    </label>
  );
};

export default TextField;
