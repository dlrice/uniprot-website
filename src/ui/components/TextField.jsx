import React from 'react';

const dataTypes = { string: 'text', integer: 'number' };

const TextField = ({
  field, type, handleChange, value = '',
}) => (
  <div className="advanced-search__inputs" key={field.term}>
    <label htmlFor={`input_${field.term}`}>
      {field.label}
      <input
        id={`input_${field.term}`}
        type={dataTypes[type]}
        value={value}
        onChange={e => handleChange(e.target.value)}
        placeholder={field.example}
      />
    </label>
  </div>
);

export default TextField;
