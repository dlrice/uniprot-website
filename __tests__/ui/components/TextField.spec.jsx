import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TextField from '../../../src/ui/components/TextField';

configure({ adapter: new Adapter() });
let props;
let wrapper;

describe('Range field', () => {
  beforeEach(() => {
    props = {
      field: {
        label: 'UniProtKB AC',
        itemType: 'single',
        term: 'accession',
        description: 'Search by UniProtKB Accession',
        example: 'P12345',
      },
      type: 'string',
      queryInput: {},
      handleChange: jest.fn(),
    };
    wrapper = shallow(<TextField {...props} />);
  });

  test('should render a text field', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should handle change', () => {
    wrapper.find(`#input_${props.field.term}`).simulate('change', { target: { value: 'val1' } });
    expect(props.handleChange).toHaveBeenCalled();
  });
});
