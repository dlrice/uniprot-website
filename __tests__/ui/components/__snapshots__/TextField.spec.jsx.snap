// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range field should render a text field 1`] = `
"<div className=\\"advanced-search__inputs\\">
  <label htmlFor=\\"input_accession\\">
    UniProtKB AC
    <input id=\\"input_accession\\" type=\\"text\\" value=\\"\\" onChange={[Function: onChange]} placeholder=\\"P12345\\" />
  </label>
</div>"
`;
